test/e2e/apimachinery/apiserver_identity.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apimachinery
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2enode "k8s.io/kubernetes/test/e2e/framework/node"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	e2essh "k8s.io/kubernetes/test/e2e/framework/ssh"
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+func getControlPlaneHostname(node *v1.Node) (string, error) {
+	nodeAddresses := e2enode.GetAddresses(node, v1.NodeExternalIP)
+	if len(nodeAddresses) == 0 {
+		return "", errors.New("no valid addresses to use for SSH")
+	}
+
+	controlPlaneAddress := nodeAddresses[0]
+
+	host := controlPlaneAddress + ":" + e2essh.SSHPort
+	result, err := e2essh.SSH("hostname", host, framework.TestContext.Provider)
+	if err != nil {
+		return "", err
+	}
+
+	if result.Code != 0 {
+		return "", fmt.Errorf("encountered non-zero exit code when running hostname command: %d", result.Code)
+	}
+
+	return result.Stdout, nil
+}
+
+// restartAPIServer attempts to restart the kube-apiserver on a node
+func restartAPIServer(node *v1.Node) error {
+	nodeAddresses := e2enode.GetAddresses(node, v1.NodeExternalIP)
+	if len(nodeAddresses) == 0 {
+		return errors.New("no valid addresses to use for SSH")
+	}
+
+	controlPlaneAddress := nodeAddresses[0]
+	cmd := "pidof kube-apiserver | xargs sudo kill"
+	framework.Logf("Restarting kube-apiserver via ssh, running: %v", cmd)
+	result, err := e2essh.SSH(cmd, net.JoinHostPort(controlPlaneAddress, e2essh.SSHPort), framework.TestContext.Provider)
+	if err != nil || result.Code != 0 {
+		e2essh.LogResult(result)
+		return fmt.Errorf("couldn't restart controller-manager: %v", err)
+	}
+	return nil
+}
+
+// This test requires that --feature-gates=APIServerIdentity=true be set on the apiserver
+var _ = SIGDescribe("kube-apiserver identity [Feature:APIServerIdentity]", func() {
+	f := framework.NewDefaultFramework("kube-apiserver-identity")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
+
+	ginkgo.It("kube-apiserver identity should persist after restart [Disruptive]", func() {
+		e2eskipper.SkipUnlessProviderIs("gce")
```

Comment:
Is this something we do often?

Reply:
Yes, especially on tests where you have to ssh into the nodes or run some external operation against them